src/transport_portals4.c

```diff
@@ -120,25 +115,184 @@ static ptl_pt_index_t heap_pt = PTL_PT_ANY;
 #endif
 
 #ifdef ENABLE_THREADS
+shmem_internal_mutex_t shmem_internal_mutex_ptl4_ctx;
 shmem_internal_mutex_t shmem_internal_mutex_ptl4_pt_state;
 shmem_internal_mutex_t shmem_internal_mutex_ptl4_frag;
 shmem_internal_mutex_t shmem_internal_mutex_ptl4_event_slots;
 shmem_internal_mutex_t shmem_internal_mutex_ptl4_nb_fence;
 #endif
 
+static shmem_transport_ctx_t** shmem_transport_portals4_contexts = NULL;
+static size_t shmem_transport_portals4_contexts_len = 0;
+static size_t shmem_transport_portals4_grow_size = 128;
+
+#define SHMEM_TRANSPORT_CTX_DEFAULT_ID -1
 shmem_transport_ctx_t shmem_transport_ctx_default;
 shmem_ctx_t SHMEM_CTX_DEFAULT = (shmem_ctx_t) &shmem_transport_ctx_default;
 
-int shmem_transport_ctx_create(long options, shmem_transport_ctx_t **ctx) {
-  *ctx = NULL;
-  return 0;
+static int
+shmem_transport_ctx_init(shmem_transport_ctx_t *ctx, long options, int id)
+{
+    int ret;
+    ptl_md_t md;
+
+    ctx->id = id;
+    ctx->options = options;
+    ctx->put_md = PTL_INVALID_HANDLE;
+    ctx->get_md = PTL_INVALID_HANDLE;
+    ctx->put_ct = PTL_INVALID_HANDLE;
+    ctx->get_ct = PTL_INVALID_HANDLE;
+    shmem_internal_atomic_write(&ctx->pending_put_cntr, 0);
+    shmem_internal_atomic_write(&ctx->pending_get_cntr, 0);
+
+    /* Allocate put completion tracking resources */
+    ret = PtlCTAlloc(shmem_transport_portals4_ni_h, &ctx->put_ct);
```

Comment:
Quick note -- Portals 4 indicates that this operation (and ```PtlMDBind``` below) should thread safe, and the reference implementation also appears to support this semantic.  /crosses fingers/

No such luck in OFI.  We wrap a mutex around the context resource allocation..

Reply:
@regrant Is this correct?  Are ```PtlCTAlloc``` and ```PtlMDBind``` thread safe?

Reply:
@jdinan PtlMDBind and PtlCTAlloc should both be thread safe, and are such in the implementation. You should be good.